Shorten error messages displayed in popups

edoardopirovano opened the pull request:
This addresses #724 by running a few regexes on error messages before they are displayed in pop-ups to try to shorten them by removing elements such as log lines and stack traces that will be too lengthy to show meaningfully in a popup.

aeisenberg commented on extensions/ql-vscode/src/helpers.ts:
```diff
@@ -29,8 +29,34 @@ export async function showAndLogErrorMessage(message: string, {
+ * @param errorMessage The error message to shorten.
+ * @return A shortened version of the error message.
+ */
+function shortenErrorMessage(errorMessage: string): string {
```
Could you add some unit tests for this?

aeisenberg commented on extensions/ql-vscode/src/helpers.ts:
```diff
@@ -29,8 +29,34 @@ export async function showAndLogErrorMessage(message: string, {
+ */
+function shortenErrorMessage(errorMessage: string): string {
+  // Filter out lines corresponding to Java stack traces.
+  const stackTraceLine = /\n\s*(com|org|net)\..*\.java:[0-9]+\)(?=\n)/g;
```
Would this be better so you're not relying on newlines, which may not work on windows.

```suggestion
  const stackTraceLine = /^\s*(com|org|net)\..*\.java:[0-9]+\)$/g;
```

aeisenberg commented on extensions/ql-vscode/src/helpers.ts:
```diff
@@ -29,8 +29,34 @@ export async function showAndLogErrorMessage(message: string, {
+  let succintMessage = errorMessage.toString().replace(stackTraceLine, '');
+
+  // Filter out lines corresponding to log lines.
+  const logLine = /\n\[[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9] [0-9][0-9]:[0-9][0-9]:[0-9][0-9]\].*/g;
```
```suggestion
  const logLine = /^\[[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9] [0-9][0-9]:[0-9][0-9]:[0-9][0-9]\].*/g;
```

aeisenberg commented on extensions/ql-vscode/src/helpers.ts:
```diff
@@ -29,8 +29,34 @@ export async function showAndLogErrorMessage(message: string, {
+  succintMessage = succintMessage.replace(logLine, '');
+
+  // Trim indentation at the start of lines.
+  const indentedLine = /\n\s+/g;
```
```suggestion
  const indentedLine = /^\s+/g;
```

aeisenberg commented on extensions/ql-vscode/src/helpers.ts:
```diff
@@ -29,8 +29,34 @@ export async function showAndLogErrorMessage(message: string, {
+  succintMessage = succintMessage.replace(indentedLine, '\n');
+
+  // Remove duplicated lines.
+  const duplicatedLine = /\n(.*)\n\1/g;
```
Careful here. The `\n` won't work on windows. You should probably be working on this line-by-line instead of running the regex over each line.

aeisenberg reviewed (changes requested):
Searching for `\n` will not work on windows. It might be easier to split line-by-line and run the regex on each line separately.

Also, now that I think about it, maybe we only need to show the first line or two of the error messages (ie- only show up to, but excluding, where the stack trace starts). This would make the code much simpler.

edoardopirovano commented:
Hmm, I agree that in the specific case of the error message that started this PR looking at the first two lines would achieve the desired result. However, I think that trying to remove things we don't want from the message while preserving it as much as possible seems better in general, since there might be cases where an important bit of information on a later line. I've fixed the regexes to work on Windows line endings, and added unit tests that run with both types of line endings (they'll still break on the line endings from pre-2001 Mac OS, but I imagine we don't support this).

aeisenberg commented on extensions/ql-vscode/test/pure-tests/helpers-pure.test.ts:
```diff
@@ -19,4 +19,38 @@ describe('helpers-pure', () => {
+      runTest(input, expected);
+    });
+
+    it('removes log lines lines', async () => {
```
```suggestion
    it('removes log lines', async () => {
```

aeisenberg commented on extensions/ql-vscode/test/pure-tests/helpers-pure.test.ts:
```diff
@@ -19,4 +19,38 @@ describe('helpers-pure', () => {
+    });
+
+    it('removes duplicate lines', async () => {
+      const input = ['first line', 'last line', 'last line'];
```
What's the behaviour here:
```suggestion
      const input = ['first line', 'last line', 'last line', 'last line'];
```

Looking at the regex, I'm not sure. Could you add a unit test for that?

aeisenberg reviewed (changes requested):
Can you also add a test for the sample error message from #724?

I do have some concern that this solution is complex and fragile. All of our error messages have their important pieces in the first line or two. This solution is probably fine since it's unlikely that we will be changing formats. I'd just like to see some more tests as described.

edoardopirovano commented:
I've added additional tests per your comments, but on further reflection I think I actually do agree that this is maybe a little fragile and simply truncating the message after two lines is the easier way to address this. Happy for you to close this PR and I'll do that instead if that's still your thinking!

aeisenberg commented:
Sure. That sounds like a better plan.
